### ISSUE TITLE
GetExternalLoginInfoAsync() null when ClaimType of NameIdentifier is not added to AuthenticatedContext ClaimsIdentity

### ISSUE THREAD

**Issue by suj87** [TerribleDev/OwinOAuthProviders#96]
Not sure if this is a bug, or just my lack of understanding - I'm going to assume it's the latter - but this did trip me up for hours.... In which case, I hope someone can explain the workings to me.

In the `_Provider_AuthenticatedContext` classes (replace `_Provider_` with whichever you're using: Yahoo/Steam/whatever) - various fields are set, one of these usually corresponds to some sort of user Id e.g. in the case of the _GitHub_ provider, the field is `string Id`, and the value searched for in `JObect user` is `id`. The field is attempted to be set like this: `Id = TryGetValue(user, "id");` and so on.

Back in the `AuthenticateCoreAsync()` method (of the `_Provider_AuthenticationHandler` class) after a new `ClaimsIdentity` is set in `context.Identity` (and the mentioned fields have been given values where possible), the first claim to be set is:

`if (!string.IsNullOrEmpty(context.Id))
{
context.Identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, context.Id, XmlSchemaString, Options.AuthenticationType));
}`

Issue(?): I discovered after a few hours that if this **does not get set**, despite a new `AuthenticationTicket` being successfully returned (and other ClaimTypes subsquently being set successfully), when the app eventually refers to the `ExternalLoginCallback` controller, `GetExternalLoginInfoAsync()` always returns null.

Why is this? Is it a bug, or just lack of understanding on my part? (In which case - could someone explain why it's required?)

If the `NameIdentifier` ClaimType is required to be set, why then is the code written in such a way that this can be skipped over if a value isn't provided by the external provider? Is this a security thing - because we always expect some sort of Id to be sent back by the external provider?

If it SHOULD work without `NameIdentifier` being set, would it be better to create some generic unique user Id when one isn't provided?

**Comment by jerriep** [comment]
There are 2 things at play here... The first is the OAuth providers, and the second is ASP.NET Identity.

The `NameIdentifier` claim type is used to refer the to "primary key" or ID of an account on the relevant service. So for Facebook this will be an integer value with the Id of the user's account on FB. After a user has been authenticated, the OAuth provider will retrieve the user's information and set the correct `NameIdentifier` claim type.

Now when using ASP.NET Identity in conjuction with the OAuth providers, ASP.NET Identity will use this specific claim to store in the AspNetUserLogins table (I explained a little bit of this in this blog post: http://www.jerriepelser.com/blog/using-aspnet-oauth-providers-without-identity). Every time a user logs in with an OAuth account (their FB account for example), ASP.NET Identity will look up the value of the `NameIdentifier` claim type to look up the user.

So in case where you use ASP.NET Identity this is absolutely requited otherwise things will break down as you have rightly mentioned.

Now as the the question of whether the `NameIdentifier` claim type SHOULD be specified...? Well as I said above, when used with ASP.NET Identity, yes it should. But the thing is that the OAuth providers can also be used "standalone" without ASP.NET Identity, and in this case the `NameIdentifier` claim type is really not required. If you simply want to use it to obtain an access token, as I demonstrated in the blog post I linked above, then it is not required.

Also depending on the Scope which the user specifies, and depending on the OAuth service, the call which gets the user information _may_ fail or _may_ not return a value for the field which is used as the `NameIdentifier` claim type, and therefore there is an `if` statement to check whether the value exisits.

In most normal circumstances however this claim type should (and must) be present... Hope this makes sense. Let me know if anything needs more clarification.
